src/looprestoration.rs

```diff
@@ -23,3 +29,288 @@ pub union LooprestorationParams {
+                        #[cfg(target_arch = "x86_64")]
+                        FnVersion::AVX2 => extern_fn!([<dav1d_wiener $name _8bpc_avx2>]),
+                        #[cfg(target_arch = "x86_64")]
+                        FnVersion::AVX512ICL => extern_fn!(
+                            // With VNNI we don't need a 5-tap version.
+                            dav1d_wiener_filter7_8bpc_avx512icl
+                        ),
+                        #[cfg(any(target_arch = "arm", target_arch = "aarch64"))]
+                        FnVersion::Neon => extern_fn!([<dav1d_wiener $name _8bpc_neon>]),
+                    }
+                }
+            }};
+        }
+
+        macro_rules! sgr_asm {
+            ($asm:expr, $name:ident) => {{
+                paste::paste! {
+                    match $asm {
+                        FnVersion::Rust => call_fn!([<sgr $name _c>]),
+
+                        #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+                        FnVersion::SSSE3 => extern_fn!([<dav1d_sgr_filter $name _8bpc_ssse3>]),
+                        #[cfg(target_arch = "x86_64")]
+                        FnVersion::AVX2 => extern_fn!([<dav1d_sgr_filter $name _8bpc_avx2>]),
+                        #[cfg(target_arch = "x86_64")]
+                        FnVersion::AVX512ICL => extern_fn!([<dav1d_sgr_filter $name _8bpc_avx512icl>]),
+                        #[cfg(any(target_arch = "arm", target_arch = "aarch64"))]
+                        FnVersion::Neon => call_fn!(args, [<sgr_filter $name _neon>]),
+
+                        #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+                        FnVersion::SSE2 => unreachable!(),
+                    }
+                }
+            }};
+        }
+
+        match filter {
+            Wiener7 => wiener_asm!(version, _filter7),
+            Wiener5 => wiener_asm!(version, _filter5),
+            Sgr5x5 => sgr_asm!(version, _5x5),
+            Sgr3x3 => sgr_asm!(version, _3x3),
+            SgrMix => sgr_asm!(version, _mix),
+        }
+    }
+}
+
+impl BitDepthFnLoopRestoration for BitDepth16 {
+    unsafe fn call(
+        &self,
+        version: FnVersion,
+        filter: FilterFn,
+        p: *mut Self::Pixel,
+        stride: ptrdiff_t,
+        left: *const [Self::Pixel; 4],
+        lpf: *const Self::Pixel,
+        w: libc::c_int,
+        h: libc::c_int,
+        params: *const LooprestorationParams,
+        edges: LrEdgeFlags,
+    ) {
+        use crate::src::looprestoration_tmpl_16::{sgr_3x3_c, sgr_5x5_c, sgr_mix_c, wiener_c};
+        use FilterFn::*;
+
+        macro_rules! call_fn {
+            ($name:ident) => {
+                $name(
+                    p,
+                    stride,
+                    left,
+                    lpf,
+                    w,
+                    h,
+                    params,
+                    edges,
+                    self.bitdepth_max().into(),
+                )
+            };
+        }
```

Comment:
Oh, you're allowed to do this?  I had thought you can't due to macro hygiene, but if you can, that'd be great.  Is it because the macro is defined inside those variables' scope already, so it's like a closure?

Reply:
Yeah, I'm pretty sure it only works because the macro is defined within the fn, so it effectively "captures" its environment kinda like a closure does. Funnily enough, I only even noticed that this works because you accidentally forgot a `$` in one of your macros in `mc.rs`, but the code still compiles because there's a variable named `args` in scope.

Reply:
> Funnily enough, I only even noticed that this works because you accidentally forgot a `$` in one of your macros in `mc.rs`, but the code still compiles because there's a variable named `args` in scope.

Oh, oops!

I might use this approach then.  I was purposefully passing `args` through everything since I thought I couldn't capture it.